Tools/uorb_graph/create.py

```diff
@@ -286,8 +286,8 @@ def build(self, src_path_list, excluded_paths=[], use_topic_pubsub_union=True):
             self._publications.filter_modules(self._module_whitelist)
 
         # modules & topics sets
-        self._modules = set(self._publications.pubsubs.keys() +
-                self._subscriptions.pubsubs.keys())
+        self._modules = set(list(self._publications.pubsubs.keys()) +
+                list(self._subscriptions.pubsubs.keys()))
```

Comment:
Do you understand why/how this changed with Python 3? Just curious...

Reply:
## 1 file() -> open()
#Python2
```python
>>> help(file)
class file(object)
 |  file(name[, mode[, buffering]]) -> file object
 |  
 |  Open a file.  The mode can be 'r', 'w' or 'a' for reading (default),
 |  writing or appending.  The file will be created if it doesn't exist
 |  when opened for writing or appending; it will be truncated when
 |  opened for writing.  Add a 'b' to the mode for binary files.
 |  Add a '+' to the mode to allow simultaneous reading and writing.
 |  If the buffering argument is given, 0 means unbuffered, 1 means line
 |  buffered, and larger numbers specify the buffer size.  The preferred way
 |  to open a file is with the builtin open() function.
....
```
#Python3
```python
>>> help(file)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'file' is not defined
```
As help doc in the `Python2`, the perferred function is open, and `Python3` remove the `file` function.

## 2 dict_obj_A.keys() + dict_obj_B.keys() ->  list(dict_obj_A.keys()) + list(dict_obj_B.keys())
#Python2
```python
>>> a = {'a':1, 'b':2}
>>> a.keys + a.keys()
['a', 'b', 'a', 'b']
```
#Python3
```python
>>> a = {'a':1, 'b':2}
>>> a.keys()
dict_keys(['b', 'a'])
>>> a.keys() + a.keys()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'dict_keys' and 'dict_keys'
>>> list(a.keys()) + list(a.keys())
['b', 'a', 'b', 'a']
```
Due to the space cost of creating list object directly, especially for really big object, the Python3 return a generator for keys() method, which don't support `__add__` method. Therefore, `+` operation is invalid in Python3 in `dict_obj_A.keys() + dict_obj_B.keys`. To solve it, manually conversion is needed.

Reply:
Got it! Thanks a lot for the information.